File: javascript/src/functions/dispatch.js
// <copyright file="dispatch.js" company="Microsoft Corporation">
// Copyright (c) Microsoft Corporation. All rights reserved.
// </copyright>

'use strict';

const crypto = require('crypto');
const { compactDecrypt } = require('jose');
const { readConfig } = require('./config');
const { DeliveryContext, TextToVoice } = require('./models');
const { providerCredentials, reportRefreshFailure } = require('./credentials');

const CHANNEL_BY_CODE = Object.freeze({ 1: 'sms', 2: 'voice' });
const CHANNEL_BY_NAME = Object.freeze({ sms: 1, voice: 2 });
const MODE = Object.freeze({ LIVE: 1, EVALUATION: 2 });
const MODE_BY_NAME = Object.freeze({ live: 1, evaluation: 2 });

function normalizeChannel(channel) {
    if (channel === 1 || channel === 2) return channel;
    if (typeof channel === 'string' && Object.hasOwn(CHANNEL_BY_NAME, channel.toLowerCase())) {
        return CHANNEL_BY_NAME[channel.toLowerCase()];
    }
    return null;
}
function normalizeMode(mode) {
    if (mode === MODE.LIVE || mode === MODE.EVALUATION) return mode;
    if (typeof mode === 'string' && Object.hasOwn(MODE_BY_NAME, mode.toLowerCase())) {
        return MODE_BY_NAME[mode.toLowerCase()];
    }
    return null;
}

/** @returns {{envelope?: import('./models').Envelope, error?: string}} */
function parseEnvelope(payload) {
    if (!payload || typeof payload !== 'object' || Array.isArray(payload)) {
        return { error: 'invalid envelope' };
    }
    const { type, tenantId, correlationId, channel, mode, ttlSeconds, encryptedDeliveryContext } = payload;
    if (type !== 'microsoft.mfa.otpDeliver.v1') {
        return { error: 'unsupported envelope type' };
    }
    if (typeof encryptedDeliveryContext !== 'string' || !encryptedDeliveryContext.trim()) {
        return { error: 'encryptedDeliveryContext is required' };
    }
    const channelCode = normalizeChannel(channel);
    if (!channelCode) {
        return { error: 'unsupported channel' };
    }
    const modeCode = normalizeMode(mode);
    if (!modeCode) {
        return { error: 'unsupported mode' };
    }
    if (Object.hasOwn(payload, 'ttlSeconds')) {
        if (!Number.isInteger(ttlSeconds) || ttlSeconds > 2147483647) {
            return { error: 'invalid ttlSeconds' };
        }
        if (ttlSeconds <= 0) {
            return { error: 'ttlSeconds expired' };
        }
    }
    return { envelope: { type, tenantId, correlationId, channel: channelCode, mode: modeCode, ttlSeconds, encryptedDeliveryContext } };
}

// Reject oversized or structurally invalid JWEs before decoding or allocating buffers.
const MAX_JWE_LENGTH = 16384;

function assertWellFormedJwe(compactJwe) {
    if (typeof compactJwe !== 'string' || compactJwe.length === 0) {
        throw new Error('malformed JWE');
    }
    if (compactJwe.length > MAX_JWE_LENGTH) {
        throw new Error('delivery context exceeds size limit');
    }
    const segments = compactJwe.split('.');
    if (segments.length !== 5 || segments.some((segment) => segment.length === 0)) {
        throw new Error('malformed JWE: expected five non-empty segments');
    }
}

function readProtectedHeader(compactJwe) {
    const protectedSegment = String(compactJwe).split('.')[0] || '';
    return JSON.parse(Buffer.from(protectedSegment, 'base64url').toString('utf8'));
}

let cachedKey;
let cachedKeyPem;

function normalizePem(value) {
    const text = String(value || '');
    if (text.includes('-----BEGIN')) return text;
    return Buffer.from(text, 'base64').toString('utf8');
}

function loadPrivateKey(pem) {
    if (!pem) {
        throw new Error('private key unavailable (EPP_DECRYPTION_KEY_PEM is not set)');
    }
    if (cachedKey && cachedKeyPem === pem) {
        return cachedKey;
    }
    cachedKey = crypto.createPrivateKey(normalizePem(pem));
    cachedKeyPem = pem;
    return cachedKey;
}

async function decryptDeliveryContext(compactJwe, config = readConfig()) {
    assertWellFormedJwe(compactJwe);
    const header = readProtectedHeader(compactJwe);
    const privateKey = loadPrivateKey(config.decryptionKeyPem);
    // Pin alg/enc so a tampered header can't downgrade the crypto.
    const { plaintext } = await compactDecrypt(compactJwe, privateKey, {
        keyManagementAlgorithms: ['RSA-OAEP-256'],
        contentEncryptionAlgorithms: ['A256GCM'],
    });
    return { header, context: DeliveryContext.fromPayload(JSON.parse(Buffer.from(plaintext).toString('utf8'))) };
}

/**
 * @param {DeliveryContext} context
 * @param {import('./models').Envelope} envelope
 * @param {string} messageId
 * @returns {import('./models').DispatchRequest}
 */
function contextToDispatch(context, envelope, messageId) {
    const channel = CHANNEL_BY_CODE[envelope.channel];
    return {
        destination: context.phoneNumber,
        message: context.message,
        channel,
        messageId,
        correlationId: envelope.correlationId,
        locale: context.locale || undefined,
        textToVoice: context.textToVoice,
    };
}

const OUTCOME = Object.freeze({
    CONTINUE: 'Continue',
    FAIL: 'Fail',
    BLOCK: 'Block',
    STEP_UP: 'StepUp',
});

const providerRegistry = new Map(
    [
        require('./providers/infobip'),
        require('./providers/sinch'),
        require('./providers/soprano'),
        require('./providers/telesign'),
    ].map((providerModule) => [
        providerModule.manifest.id.toLowerCase(),
        { manifest: providerModule.manifest, adapter: providerModule },
    ]),
);

function getProvider(providerId) {
    return providerId ? providerRegistry.get(String(providerId).trim().toLowerCase()) || null : null;
}

async function resolveProviderCredential(authConfiguration = {}, config) {
    return providerCredentials.resolve(authConfiguration, config);
}

async function startProviderCredentialRefresh() {
    const config = readConfig();
    if (!config.providerName) return;
    const provider = getProvider(config.providerName);
    if (!provider || (config.providerAuthMode && config.providerAuthMode !== provider.manifest.auth.mode)) {
        reportRefreshFailure('configuration');
        return;
    }
    try {
        await resolveProviderCredential(provider.manifest.auth, config);
    } catch {
        // The cache reports acquisition failures; also report configurations rejected before caching.
        if (!providerCredentials.current) reportRefreshFailure('configuration');
    }
}

function stopProviderCredentialRefresh() {
    providerCredentials.close();
}

// Status mappings may restrict HTTP success, but cannot turn failed HTTP into Continue.
/** @param {import('./models').ParsedResponse} parsedResponse */
function resolveOutcome(manifest, parsedResponse) {
    const responseMapping = manifest.responseMapping || {};
    const providerStatusKey = parsedResponse.providerStatusName || parsedResponse.providerStatusCode;
    const fallback = Object.hasOwn(responseMapping, 'default')
        ? responseMapping.default || OUTCOME.FAIL : OUTCOME.FAIL;
    const hasMapping = (typeof providerStatusKey === 'string' || typeof providerStatusKey === 'number')
        && Object.hasOwn(responseMapping, providerStatusKey);
    const outcome = providerStatusKey
        ? (hasMapping ? responseMapping[providerStatusKey] || fallback : fallback)
        : (parsedResponse.success ? OUTCOME.CONTINUE : fallback);
    return outcome === OUTCOME.CONTINUE && !parsedResponse.success ? OUTCOME.FAIL : outcome;
}

function outcomeToHttpStatus(outcome, providerHttpStatus) {
    switch (outcome) {
        case OUTCOME.CONTINUE:
            return 200;
        case OUTCOME.BLOCK:
            return 403;
        case OUTCOME.STEP_UP:
            return 409;
        case OUTCOME.FAIL:
            if (providerHttpStatus === 429) return 429;
            if (providerHttpStatus === 401 || providerHttpStatus === 403) return 401;
            if (providerHttpStatus >= 400 && providerHttpStatus < 500) return 400;
            return 502;
        default:
            return 502;
    }
}

// App settings use one grammar: trimmed ASCII decimal digits, no sign, exponent, or hex.
function parseProviderTimeout(value) {
    const text = typeof value === 'string' ? value.trim() : '';
    if (!text || [...text].some((character) => character < '0' || character > '9')) return 1500;
    const milliseconds = Number(text);
    return milliseconds > 0 ? Math.min(milliseconds, 2500) : 1500;
}

function isValidProviderUrl(value) {
    if (typeof value !== 'string' || !value.toLowerCase().startsWith('https://')) return false;
    for (const character of value) {
        if (!character.trim() || character.charCodeAt(0) < 32 || character === '\\' || character === '#') return false;
    }
    const authority = value.slice('https://'.length).split('/')[0].split('?')[0];
    // URL normalizes empty userinfo and empty ports away; reject those in the original authority too.
    if (!authority || authority.includes('@') || authority.endsWith(':')) return false;
    try {
        const url = new URL(value);
        return url.protocol === 'https:' && !!url.hostname && !url.username && !url.password && !url.hash
            && (!url.port || (Number(url.port) >= 1 && Number(url.port) <= 65535));
    } catch {
        return false;
    }
}

async function fetchWithTimeout(providerRequest, timeoutMilliseconds, log) {
    const abortController = new AbortController();
    let timedOut = false;
    const timeoutTimer = setTimeout(() => {
        timedOut = true;
        abortController.abort();
    }, timeoutMilliseconds);

    try {
        log?.providerRequestStarted(timeoutMilliseconds);
        const response = await fetch(providerRequest.url, {
            method: providerRequest.method || 'POST',
            headers: providerRequest.headers,
            body: providerRequest.body,
            signal: abortController.signal,
            redirect: 'manual', // Never forward provider credentials to a redirect target.
        });
        log?.providerResponseReceived(response.status);
        const responseText = await response.text();
        return { response, responseText };
    } catch {
        const error = new Error('provider request failed');
        error.name = timedOut ? 'TimeoutError' : 'Error';
        throw error;
    } finally {
        clearTimeout(timeoutTimer);
        log?.providerRequestFinished();
    }
}

const failBody = (providerId, channel, reason, dispatch, requestId) =>
    ({ status: 'failed', outcome: OUTCOME.FAIL, provider: providerId, channel, reason, correlationId: dispatch.correlationId, messageId: dispatch.messageId, requestId });

async function sendViaProvider(providerEntry, dispatch, options) {
    const { requestId, config, log } = options;
    const { manifest, adapter } = providerEntry;
    const providerId = manifest.id;
    const channel = dispatch.channel === undefined ? 'sms'
        : (typeof dispatch.channel === 'string' ? dispatch.channel.toLowerCase() : null);

    if (!['sms', 'voice'].includes(channel)) {
        log?.failure('provider_configuration', 'unsupported_channel', 400);
        return { httpStatus: 400, body: { status: 'error', reason: 'unsupported channel', requestId } };
    }
    if (config.providerChannel && config.providerChannel !== channel) {
        log?.failure('provider_configuration', 'channel_not_configured', 400);
        return { httpStatus: 400, body: { status: 'error', provider: providerId, reason: 'channel not configured', requestId } };
    }
    if (config.providerAuthMode && config.providerAuthMode !== manifest.auth?.mode) {
        log?.failure('provider_configuration', 'authentication_mode_mismatch', 502);
        return { httpStatus: 502, body: failBody(providerId, channel, 'provider authentication mismatch', dispatch, requestId) };
    }

    if (channel === 'voice' && manifest.requiresTextToVoice
        && (!(dispatch.textToVoice instanceof TextToVoice) || !dispatch.textToVoice.isComplete)) {
        log?.failure('provider_request_build', 'incomplete_voice_context', 400);
        return { httpStatus: 400, body: failBody(providerId, channel, 'incomplete voice context', dispatch, requestId) };
    }

    const endpointBaseUrl = config.providerEndpoint;
    if (!isValidProviderUrl(endpointBaseUrl)) {
        log?.failure('provider_configuration', 'invalid_provider_endpoint', 502);
        return { httpStatus: 502, body: failBody(providerId, channel, 'provider endpoint missing or invalid', dispatch, requestId) };
    }

    let credential = null;
    try {
        log?.credentialResolutionStarted(config);
        credential = await resolveProviderCredential(manifest.auth, config);
    } catch {
        log?.failure('provider_credentials', 'credential_unavailable', 502);
        return { httpStatus: 502, body: failBody(providerId, channel, 'provider credential unavailable', dispatch, requestId) };
    }
    const identityRequired = credential?.mode === 'apiKey' && !!manifest.auth?.identityKeyVaultSecretName;
    const credentialUnavailable = !credential
        || (credential.mode === 'apiKey' && (!credential.secret || (identityRequired && !credential.identity)))
        || (credential.mode === 'oauth' && !credential.accessToken);
    if (credentialUnavailable) {
        log?.failure('provider_credentials', 'credential_unavailable', 502);
        return { httpStatus: 502, body: failBody(providerId, channel, 'provider credential unavailable', dispatch, requestId) };
    }
    log?.credentialResolved();

    let providerRequest;
    try {
        log?.service('provider_request_build_started');
        providerRequest = adapter.buildRequest({
            channel,
            endpoint: endpointBaseUrl,
            dispatch,
            credential,
            env: config.env,
        });
    } catch {
        log?.failure('provider_request_build', 'request_build_failed', 502);
        return { httpStatus: 502, body: failBody(providerId, channel, 'provider request failed', dispatch, requestId) };
    }

    if (!isValidProviderUrl(providerRequest.url)) {
        log?.failure('provider_request_build', 'invalid_provider_request_url', 502);
        return { httpStatus: 502, body: failBody(providerId, channel, 'provider request URL invalid', dispatch, requestId) };
    }
    log?.providerRequestBuilt(providerRequest.method || 'POST', providerRequest.url);

    const timeoutMilliseconds = parseProviderTimeout(config.providerTimeoutMs);
    let providerResponse;
    let responseText;
    try {
        ({ response: providerResponse, responseText } = await fetchWithTimeout(providerRequest, timeoutMilliseconds, log));
    } catch (error) {
        const isTimeout = error.name === 'TimeoutError';
        const httpStatus = isTimeout ? 504 : 502;
        log?.failure('provider_transport', isTimeout ? 'provider_timeout' : 'provider_network_error', httpStatus);
        return { httpStatus, body: failBody(providerId, channel, isTimeout ? 'provider request timed out' : 'provider request failed', dispatch, requestId) };
    }

    let responseJson;
    let validJson = true;
    try {
        responseJson = JSON.parse(responseText);
    } catch {
        validJson = false;
        log?.service('provider_response_invalid_json', {}, 'warn');
        responseJson = {};
    }

    let parsedResponse;
    let outcome;
    let httpStatus;
    try {
        parsedResponse = adapter.parseResponse({
            httpStatus: providerResponse.status,
            ok: providerResponse.ok,
            json: responseJson,
        });
        outcome = resolveOutcome(manifest, parsedResponse);
        httpStatus = outcomeToHttpStatus(outcome, parsedResponse.providerHttpStatus);
    } catch (error) {
        log?.failure('provider_response', 'response_parse_failed', 500);
        throw error;
    }
    log?.providerResponseProcessed(manifest, parsedResponse, outcome, httpStatus, validJson);

    return {
        httpStatus,
        body: {
            status: outcome === OUTCOME.CONTINUE ? 'accepted' : 'failed',
            outcome,
            provider: providerId,
            channel,
            messageId: dispatch.messageId,
            correlationId: dispatch.correlationId,
            requestId,
        },
    };
}

async function dispatchOtp(dispatch, { config = readConfig(), requestId, log } = {}) {
    const providerEntry = getProvider(config.providerName);
    if (!providerEntry) {
        log?.failure('provider_selection', 'unknown_provider', 400);
        return {
            httpStatus: 400,
            body: { status: 'error', reason: 'unknown provider', requestId },
        };
    }
    log?.providerSelected(providerEntry.manifest);
    return sendViaProvider(providerEntry, dispatch, { config, requestId, log });
}

module.exports = {
    parseEnvelope,
    decryptDeliveryContext,
    contextToDispatch,
    MODE,
    dispatchOtp,
    getProvider,
    resolveOutcome,
    outcomeToHttpStatus,
    parseProviderTimeout,
    isValidProviderUrl,
    resolveProviderCredential,
    startProviderCredentialRefresh,
    stopProviderCredentialRefresh,
};

File: python/src/dispatch.py
import base64
import json
import os
from urllib.parse import urlsplit

import requests
from jwcrypto import jwe as jwe_module
from jwcrypto import jwk
from urllib3.exceptions import ReadTimeoutError

from .config import read_config
from .models import DeliveryContext, DispatchRequest, Envelope

DEFAULT_TIMEOUT_MS = 1500
DEFAULT_CHANNELS = ["sms", "voice"]

CONTINUE = "Continue"
FAIL = "Fail"
BLOCK = "Block"
STEP_UP = "StepUp"


def resolve_outcome(manifest, parsed):
    mapping = manifest["response_mapping"]
    key = parsed.get("provider_status_name") or parsed.get("provider_status_code")
    if key:
        outcome = mapping.get(key) or mapping.get("default", FAIL)
    else:
        outcome = CONTINUE if parsed.get("success") else mapping.get("default", FAIL)
    return FAIL if outcome == CONTINUE and not parsed.get("success") else outcome


def to_http_status(outcome, provider_http_status):
    if outcome == CONTINUE:
        return 200
    if outcome == BLOCK:
        return 403
    if outcome == STEP_UP:
        return 409
    if outcome == FAIL:
        if provider_http_status == 429:
            return 429
        if provider_http_status in (401, 403):
            return 401
        if 400 <= provider_http_status < 500:
            return 400
    return 502


class ProviderRegistry:
    def __init__(self, adapters):
        self._by_id = {adapter.manifest["id"].lower(): adapter for adapter in adapters}

    def get(self, provider_id):
        if not provider_id:
            return None
        return self._by_id.get(provider_id.lower())


CHANNEL_BY_CODE = {1: "sms", 2: "voice"}
CHANNEL_BY_NAME = {"sms": 1, "voice": 2}
MODE_LIVE = 1
MODE_EVALUATION = 2
MODE_BY_NAME = {"live": MODE_LIVE, "evaluation": MODE_EVALUATION}

MAX_JWE_LENGTH = 16384


def _normalize_channel(channel):
    if type(channel) is int:
        return channel if channel in CHANNEL_BY_CODE else None
    if isinstance(channel, str):
        return CHANNEL_BY_NAME.get(channel.lower())
    return None


def _normalize_mode(mode):
    if type(mode) is int:
        return mode if mode in (MODE_LIVE, MODE_EVALUATION) else None
    if isinstance(mode, str):
        return MODE_BY_NAME.get(mode.lower())
    return None


def parse_envelope(payload) -> tuple[Envelope | None, str | None]:
    if not isinstance(payload, dict):
        return None, "invalid envelope"
    if payload.get("type") != "microsoft.mfa.otpDeliver.v1":
        return None, "unsupported envelope type"
    encrypted = payload.get("encryptedDeliveryContext")
    if not isinstance(encrypted, str) or not encrypted.strip():
        return None, "encryptedDeliveryContext is required"
    channel = _normalize_channel(payload.get("channel"))
    if channel is None:
        return None, "unsupported channel"
    mode = _normalize_mode(payload.get("mode"))
    if mode is None:
        return None, "unsupported mode"
    ttl_seconds = payload.get("ttlSeconds")
    if "ttlSeconds" in payload:
        if type(ttl_seconds) is not int:
            return None, "invalid ttlSeconds"
        if ttl_seconds <= 0:
            return None, "ttlSeconds expired"
        if ttl_seconds > 2147483647:
            return None, "invalid ttlSeconds"
    return Envelope(
        type=payload.get("type"),
        tenant_id=payload.get("tenantId"),
        correlation_id=payload.get("correlationId"),
        channel=channel,
        mode=mode,
        ttl_seconds=ttl_seconds,
        encrypted_delivery_context=encrypted,
    ), None


def read_protected_header(compact_jwe):
    header_segment = compact_jwe.split(".")[0]
    header_segment += "=" * (-len(header_segment) % 4)
    return json.loads(base64.urlsafe_b64decode(header_segment))


def make_key_provider(env):
    def key_provider(_kid):
        return read_config(env).decryption_key_pem

    return key_provider


def _assert_well_formed_jwe(compact_jwe):
    if not isinstance(compact_jwe, str) or not compact_jwe:
        raise ValueError("malformed JWE")
    if len(compact_jwe) > MAX_JWE_LENGTH:
        raise ValueError("delivery context exceeds size limit")
    segments = compact_jwe.split(".")
    if len(segments) != 5 or not all(segments):
        raise ValueError("malformed JWE: expected five non-empty segments")


# Cache only the configured key to avoid repeated RSA imports.
_key_cache = {}


def _normalize_pem(value):
    # Base64 preserves PEM newlines in app settings.
    text = value if isinstance(value, str) else value.decode("utf-8")
    if "-----BEGIN" in text:
        return text
    return base64.b64decode(text).decode("utf-8")


def _load_private_key(pem):
    if not pem:
        raise ValueError("private key unavailable (EPP_DECRYPTION_KEY_PEM is not set)")
    cached = _key_cache.get(pem)
    if cached is None:
        cached = jwk.JWK.from_pem(_normalize_pem(pem).encode("utf-8"))
        _key_cache.clear()
        _key_cache[pem] = cached
    return cached


def decrypt_delivery_context(compact_jwe, key_provider):
    _assert_well_formed_jwe(compact_jwe)
    header = read_protected_header(compact_jwe)
    key = _load_private_key(key_provider(header.get("kid")))
    token = jwe_module.JWE(algs=["RSA-OAEP-256", "A256GCM"])
    token.deserialize(compact_jwe, key=key)
    payload = json.loads(token.payload.decode("utf-8"))
    return header, DeliveryContext.from_payload(payload)


def context_to_dispatch(context, envelope, message_id):
    return DispatchRequest(
        destination=context.phone_number,
        message=context.message,
        channel=CHANNEL_BY_CODE[envelope.channel],
        message_id=message_id,
        correlation_id=envelope.correlation_id,
        locale=context.locale,
    )


def _valid_provider_url(value):
    if not isinstance(value, str) or not value or "#" in value:
        return False
    # urlsplit strips controls; reject them before parsing.
    if any(character.isspace() or ord(character) < 32 or ord(character) == 127 for character in value):
        return False
    try:
        parsed = urlsplit(value)
        port = parsed.port  # Access validates the port's syntax and range.
        return (
            parsed.scheme == "https"
            and bool(parsed.hostname)
            and port != 0
            and parsed.username is None
            and parsed.password is None
            and not parsed.fragment
            and not parsed.netloc.endswith(":")
        )
    except ValueError:
        return False


def _provider_timeout_ms(value):
    digits = value.strip() if isinstance(value, str) else ""
    if not digits or not digits.isascii() or not digits.isdecimal():
        return DEFAULT_TIMEOUT_MS
    digits = digits.lstrip("0")
    if not digits:
        return DEFAULT_TIMEOUT_MS
    # Clamp before int() to avoid its digit limit.
    if len(digits) > 4 or (len(digits) == 4 and digits > "2500"):
        return 2500
    return int(digits)


def _has_read_timeout(error):
    # requests wraps urllib3 body-read timeouts in ConnectionError.
    pending = [error]
    seen = set()
    while pending:
        current = pending.pop()
        if id(current) in seen:
            continue
        seen.add(id(current))
        if isinstance(current, ReadTimeoutError):
            return True
        pending.extend(
            nested for nested in (current.__cause__, current.__context__, *current.args)
            if isinstance(nested, Exception)
        )
    return False


class DispatchEngine:
    def __init__(self, registry, secrets, env=None):
        self.registry = registry
        self.secrets = secrets
        self.env = env if env is not None else os.environ

    def dispatch(self, dispatch, request_id):
        config = read_config(self.env)
        adapter = self.registry.get(config.provider_name)
        if adapter is None:
            return 400, {"status": "error", "reason": "unknown provider", "requestId": request_id}

        manifest = adapter.manifest
        provider_id = manifest["id"]
        channel = dispatch.channel if dispatch.channel is not None else "sms"
        if not isinstance(channel, str):
            return 400, {"status": "error", "provider": provider_id, "reason": "unsupported channel", "requestId": request_id}
        channel = channel.lower()

        if channel not in DEFAULT_CHANNELS:
            return 400, {"status": "error", "provider": provider_id, "reason": "unsupported channel", "requestId": request_id}

        auth = manifest["auth"]
        if auth.get("mode") != "apiKey":
            return 502, self._fail_body(provider_id, channel, "unsupported provider auth mode", dispatch, request_id)
        try:
            credential = self._resolve_credential(auth)
        except Exception:
            return 502, self._fail_body(provider_id, channel, "provider credential unavailable", dispatch, request_id)
        if not credential.get("secret") or (auth.get("identity_key_vault_secret_name") and not credential.get("identity")):
            return 502, self._fail_body(provider_id, channel, "provider credential unavailable", dispatch, request_id)

        endpoint = config.provider_endpoint
        if not endpoint:
            return 502, self._fail_body(provider_id, channel, "provider endpoint not configured", dispatch, request_id)
        if not _valid_provider_url(endpoint):
            return 502, self._fail_body(provider_id, channel, "invalid provider endpoint", dispatch, request_id)

        try:
            provider_request = adapter.build_request(channel, endpoint, dispatch, credential, config.env)
        except Exception:
            return 502, self._fail_body(provider_id, channel, "provider request failed", dispatch, request_id)
        if not _valid_provider_url(provider_request.get("url")):
            return 502, self._fail_body(provider_id, channel, "invalid provider request URL", dispatch, request_id)

        timeout_ms = _provider_timeout_ms(config.provider_timeout_ms)
        response = None
        try:
            response = requests.request(
                provider_request["method"],
                provider_request["url"],
                headers=provider_request["headers"],
                data=provider_request["body"],
                # Connect/read inactivity, not a total delivery deadline.
                timeout=timeout_ms / 1000,
                allow_redirects=False,  # Never forward credentials to a redirect target.
                stream=True,  # Own the response for cleanup if body reading fails.
            )

            try:
                body_json = response.json()
            except ValueError:
                body_json = {}

            ok = 200 <= response.status_code < 300
            parsed = adapter.parse_response(response.status_code, ok, body_json)
            outcome = resolve_outcome(manifest, parsed)
            http_status = to_http_status(outcome, parsed.get("provider_http_status") or response.status_code)

            return http_status, {
                "status": "accepted" if outcome == CONTINUE else "failed",
                "outcome": outcome,
                "provider": provider_id,
                "channel": channel,
                "messageId": dispatch.message_id,
                "correlationId": dispatch.correlation_id,
                "requestId": request_id,
            }
        except requests.exceptions.RequestException as error:
            if response is None:
                response = getattr(error, "response", None)
            if isinstance(error, requests.exceptions.Timeout) or (
                isinstance(error, requests.exceptions.ConnectionError) and _has_read_timeout(error)
            ):
                return 504, self._fail_body(provider_id, channel, "provider timeout", dispatch, request_id)
            return 502, self._fail_body(provider_id, channel, "provider request failed", dispatch, request_id)
        except Exception:
            return 502, self._fail_body(provider_id, channel, "provider response failed", dispatch, request_id)
        finally:
            close = getattr(response, "close", None)
            if callable(close):
                try:
                    close()
                except Exception:
                    pass

    def _resolve_credential(self, auth):
        secret = self.secrets.resolve(auth.get("key_vault_secret_name"))
        identity = self.secrets.resolve(auth.get("identity_key_vault_secret_name")) if auth.get("identity_key_vault_secret_name") else ""
        return {"mode": "apiKey", "secret": secret, "identity": identity}

    def _fail_body(self, provider, channel, reason, dispatch, request_id):
        return {"status": "failed", "outcome": "Fail", "provider": provider, "channel": channel, "reason": reason, "correlationId": dispatch.correlation_id, "messageId": dispatch.message_id, "requestId": request_id}
